spring-cloud-stream-binders/spring-cloud-stream-binder-gemfire/src/main/java/org/springframework/cloud/stream/binder/gemfire/MessageKey.java

```diff
@@ -0,0 +1,229 @@
+
+	/**
+	 * Unique sequential id for message.
+	 */
+	private long sequenceId;
+
+	/**
+	 * Encoded identifier for producer. First (leftmost) 16 bits are
+	 * reserved for the PID of the producer process, rightmost 48
+	 * bits are reserved for the timestamp of when the producer
+	 * process was started.
+	 */
+	private long producerId;
+
+	/**
+	 * Object returned for {@link #getRoutingObject}.
+	 */
+	private int routingHash;
+
+	/**
+	 * Constructor for serialization.
+	 */
+	public MessageKey() {
+	}
+
+	/**
+	 * Construct a {@code MessageKey}.
+	 *
+	 * @param sequenceId unique sequence id used for ordering messages
+	 * @param producerTimestamp time stamp of when the producer process started
+	 * @param producerPid PID of producer process
+	 */
+	public MessageKey(long sequenceId, long producerTimestamp, int producerPid) {
+		this.sequenceId = sequenceId;
+		this.producerId = producerId(producerTimestamp, producerPid);
+		this.routingHash = this.hashCode();
+	}
+
+	/**
+	 * Construct a {@code MessageKey} with a specific hash for selecting the
+	 * partitioned region bucket.
+	 *
+	 * @param sequenceId unique sequence id used for ordering messages
+	 * @param producerTimestamp time stamp of when the producer process started
+	 * @param producerPid PID of producer process
+	 * @param routingHash routing hash for bucket selection
+	 */
+	public MessageKey(long sequenceId, long producerTimestamp, int producerPid, int routingHash) {
+		this.sequenceId = sequenceId;
+		this.producerId = producerId(producerTimestamp, producerPid);
+		this.routingHash = routingHash;
+	}
+
+	/**
+	 * Return a producer id consisting of the timestamp of the process startup
+	 * time for the producer and the PID of the producer.
+	 * <p>
+	 * The first (leftmost) 16 bits are reserved for the PID of the producer
+	 * process, rightmost 48 bits are reserved for the timestamp of when the
+	 * producer process was started.
+	 *
+	 * @param producerTimestamp startup timestamp for producer
+	 * @param producerPid PID for producer
+	 * @return producer id
+	 */
+	private long producerId(long producerTimestamp, int producerPid) {
+		long longPid = producerPid;
+		longPid = longPid << 48;
+		return longPid ^ producerTimestamp;
+	}
+
+	/**
+	 * Return the sequence id used for ordering messages.
+	 *
+	 * @return sequence id
+	 */
+	public long getSequenceId() {
+		return sequenceId;
+	}
+
+	/**
+	 * Return the producer id for this message key.
+	 *
+	 * @return producer id
+	 */
+	public long getProducerId() {
+		return producerId;
+	}
+
+	/**
+	 * Return the PID of the producer for this message key.
+	 *
+	 * @return producer PID
+	 */
+	public int getProducerPid() {
+		long longPid = producerId >>> 48;
+		return (int) longPid;
+	}
+
+	/**
+	 * Return the timestamp of when the producer of this message key started.
+	 *
+	 * @return producer startup timestamp
+	 */
+	public long getProducerTimestamp() {
+		return producerId << 16 >> 16;
```

Comment:
Keep the `producerTimestamp` as a field to avoid the need to strip out the pid on every `compareTo()`.